node/impl/client/client.go

```diff
@@ -763,72 +755,74 @@ func (a *API) clientRetrieve(ctx context.Context, order api.RetrievalOrder, ref
 			}
 		})
 
-		dealID, err := a.Retrieval.Retrieve(
+		resp, err := a.Retrieval.Retrieve(
 			ctx,
 			order.Root,
 			params,
 			order.Total,
 			*order.MinerPeer,
 			order.Client,
-			order.Miner,
-			store.StoreID())
+			order.Miner)
 
 		if err != nil {
 			unsubscribe()
 			finish(xerrors.Errorf("Retrieve failed: %w", err))
 			return
 		}
 
-		err = readSubscribeEvents(ctx, dealID, subscribeEvents, events)
+		err = readSubscribeEvents(ctx, resp.DealID, subscribeEvents, events)
 
 		unsubscribe()
 		if err != nil {
 			finish(xerrors.Errorf("Retrieve: %w", err))
 			return
 		}
-	} else {
-		// local retrieval
-		st, err := ((*multistore.MultiStore)(a.Mds)).Get(*order.LocalStore)
-		if err != nil {
-			finish(xerrors.Errorf("Retrieve: %w", err))
-			return
-		}
 
-		store = &multiStoreRetrievalStore{
-			storeID: *order.LocalStore,
-			store:   st,
-		}
-	}
-
-	// If ref is nil, it only fetches the data into the configured blockstore.
-	if ref == nil {
-		finish(nil)
-		return
+		carV2FilePath = resp.CarFilePath
+		// remove the temp CARv2 file when retrieval is complete
+		defer os.Remove(carV2FilePath) //nolint:errcheck
+	} else {
+		carV2FilePath = order.LocalCARV2FilePath
 	}
 
-	rdag := store.DAGService()
-
 	if ref.IsCAR {
+		// user wants a CAR file, transform the CARv2 to a CARv1 and write it out.
 		f, err := os.OpenFile(ref.Path, os.O_CREATE|os.O_WRONLY, 0644)
 		if err != nil {
 			finish(err)
 			return
 		}
-		err = car.WriteCar(ctx, rdag, []cid.Cid{order.Root}, f)
+
+		carv2Reader, err := carv2.NewReaderMmap(carV2FilePath)
 		if err != nil {
 			finish(err)
 			return
 		}
+		defer carv2Reader.Close() //nolint:errcheck
+		if _, err := io.Copy(f, carv2Reader.CarV1Reader()); err != nil {
+			finish(err)
+			return
+		}
+
 		finish(f.Close())
 		return
 	}
 
-	nd, err := rdag.Get(ctx, order.Root)
+	rw, err := blockstore.OpenReadOnly(carV2FilePath, false)
```

Comment:
supernit: change var name from rw to readOnly?

Reply:
Done.